### PR TITLE
Stack: Promoting stack from experiments package to OUFR package.

### LINKED ISSUE #5419: New Component: Stack
## Component Details

A Stack is a container-type component that abstracts the implementation of a flexbox in order to define the layout of its children components.

#### Why should we abstract flexbox?

The reason for abstracting flexbox comes from our drive to provide a complete component library, where we wouldn't want for the consumer to need to write additional CSS just to get flexbox functionality. By abstracting flexbox we:
1. Provide an official Fabric, CSS-less way of getting flexbox functionality into the apps.
2. Remove some of the mental strain that comes from thinking about the axes when using flexboxes.

#### Stack properties

Although Stack has a number of different properties, there are three in particular that define the overall layout that the component has:

- Direction: Refers to whether the stacking of children components is horizontal or vertical. By default the Stack component is vertical, but can be turned horizontal by adding the `horizontal` property when using the component.
- Alignment: Refers to how are the children components aligned inside the container. This is controlled via the `verticalAlign` and `horizontalAlign` properties. One thing to notice here is that while flexbox containers align always across the cross axis, Stack aims to remove the mental strain involved in this process by making the `verticalAlign` and `horizontalAlign` properties always follow the vertical and horizontal axes, respectively, regardless of the direction of the Stack.
- Spacing: Refers to the space that exists between children components inside the Stack. This is controlled via the `gap` ad `verticalGap` properties.

#### Stack wrapping

Aside from the previously mentioned properties, there is another property called `wrap` that determines if items overflow the Stack container or wrap around it. The wrap property only works in the direction of the Stack, which means that the children components can still overflow in the perpendicular direction (i.e. in a Vertical Stack, items might overflow horizontally and vice versa).

#### Stack nesting

Stacks can be nested inside one another in order to be able to configure the layout of the application as desired.

#### Description of the changes

When discussing the promotion of the component from the `experiments` package into the `office-ui-fabric-react` package, it was decided that we should change the approach we were using in the `experiments` package of having two separate components, **HorizontalStack** and **VerticalStack**, and instead have a single component **Stack** with a boolean property `horizontal` that determines if the stack is horizontal or vertical, with a vertical stack being the default.

This change required modifying existing examples, tests and usage of **Stack** inside other components throughout the `experiments` package. The **HorizontalStack** and **VerticalStack** pages in the `experiments` website have also been removed and a **Stack** page has been added to the `office-ui-fabric-react` website as part of this change.

#### Focus areas to test

All the existing tests (snapshots, vr-tests, etc) for **HorizontalStack** and **VerticalStack** have been modified to use the singular **Stack** component and the new component has been tested against all of them to ensure conformance with how the previous components worked.

## TO-DOs before promoting Stack to OUFR package

- [x] ~~Consolidate horizontal and vertical stacks into one component.~~
  - [x] ~~Update component files.~~
  - [x] ~~Update example files.~~
  - [x] ~~Update snapshot tests.~~
  - [x] ~~Update vr-tests.~~
- [x] ~~Add the `wrap` functionality to vertical Stacks.~~
  - [x] ~~Add examples in documentation page.~~
  - [x] ~~Add snapshot tests.~~
  - [x] ~~Add vr-tests.~~
- [x] ~~Finish documentation on Stack component.~~
  - [x] ~~Add overview.~~
  - [x] ~~Add Do's.~~
  - [x] ~~Add Dont's.~~
- [x] ~~Slots + Tokens in Stack~~

### Imports

Stack does not make use of other components internally to be constructed.

### Exports/ Component Breakdown

Both the Stack and StackItem components are exported, as well as the Stack properties defined in `Stack.types`.

### Intended Package

Initially in `@uifabric/experiments` while we drive the changes to have the correct api, but with the intention of moving it to `office-ui-fabric-react` after that is done.

### Code mockup/example

__Basic Vertical Stack Code Sample__

```ts
<Stack gap={5}>
   <span>Default vertical stack</span>
   <Stack className={styles.root}>
      <span>Item One</span>
      <span>Item Two</span>
      <span>Item Three</span>
   </Stack>
   <span>Vertical gap between items</span>
   <Stack gap={10} padding={10} className={styles.root}>
      <span>Item One</span>
      <span>Item Two</span>
      <span>Item Three</span>
   </Stack>

   <span>Item alignments</span>
   <Stack gap={5} padding={10} className={styles.root}>
      <Stack.Item align="auto" className={styles.item}>
         <span>Auto-aligned item</span>
      </Stack.Item>
      <Stack.Item align="stretch" className={styles.item}>
         <span>Stretch-aligned item</span>
      </Stack.Item>
      <Stack.Item align="baseline" className={styles.item}>
         <span>Baseline-aligned item</span>
      </Stack.Item>
      <Stack.Item align="start" className={styles.item}>
         <span>Start-aligned item</span>
      </Stack.Item>
      <Stack.Item align="center" className={styles.item}>
         <span>Center-aligned item</span>
      </Stack.Item>
      <Stack.Item align="end" className={styles.item}>
         <span>End-aligned item</span>
      </Stack.Item>
   </Stack>

   <span>Clickable vertical stack</span>
   <Stack onClick={this._onClick} padding={10} className={styles.root}>
      <span>Click inside this box</span>
   </Stack>
</Stack>
```

image

__Basic Horizontal Stack Code Sample__

```ts
<Stack gap={5}>
   <span>Default horizontal stack</span>
   <Stack horizontal className={styles.root}>
      <span>Item One</span>
      <span>Item Two</span>
      <span>Item Three</span>
   </Stack>

   <span>Horizontal gap between items</span>
   <Stack horizontal gap={10} padding={10} className={styles.root}>
      <span>Item One</span>
      <span>Item Two</span>
      <span>Item Three</span>
   </Stack>

   <span>Item alignments</span>
   <Stack horizontal gap={5} padding={10} className={styles.root} styles={{ root: { height: 100 } }}>
      <Stack.Item align="auto" className={styles.item}>
         <span>Auto-aligned item</span>
      </Stack.Item>
      <Stack.Item align="stretch" className={styles.item}>
         <span>Stretch-aligned item</span>
      </Stack.Item>
      <Stack.Item align="baseline" className={styles.item}>
         <span>Baseline-aligned item</span>
      </Stack.Item>
      <Stack.Item align="start" className={styles.item}>
         <span>Start-aligned item</span>
      </Stack.Item>
      <Stack.Item align="center" className={styles.item}>
         <span>Center-aligned item</span>
      </Stack.Item>
      <Stack.Item align="end" className={styles.item}>
         <span>End-aligned item</span>
      </Stack.Item>
   </Stack>

   <span>Clickable stack</span>
   <Stack horizontal onClick={this._onClick} padding={10} className={styles.root}>
      <span>Click inside this box</span>
   </Stack>
</Stack>
```
image

### Stack Types

__IStackItemSlots__

| Name | Type | Description |
|--------|:------:|-------------|
| `root` | `IHTMLSpanSlot` | Defines root slot of the component |

__IStackItemProps__

| Name      | Type     | Default Value | Description |
|-----------|:---------:|:--------------:|-------------|
| `as` | `string` \| `React.ReactType<IStackProps>` |  | Defines how to render the `StackItem` |
| `className` | `string` | | Defines a CSS class name used to style the `StackItem` |
| `grow` | `boolean` \| `number` \| `inherit` \| `initial` \| `unset` | | Defines how much to grow the `StackItem` in proportion to its siblings |
| `shrink` | `boolean` \| `number` \| `inherit` \| `initial` \| `unset` | | Defines at what ratio should the `StackItem` shrink to fit the available space |
| `disableShrink` | `boolean` | `false` | Defines whether the `StackItem` should be prevented from shrinking. |
| `align` | `auto` \| `stretch` \| `baseline` \| `start` \| `center` \| `end` | | Defines how to align the `StackItem` along the x-axis (for vertical `Stacks`) or the y-axis (for horizontal `Stacks`) |
| `verticalFill` | `boolean` | `true` | Defines whether the `StackItem` should take up 100% of the height of its parent |

__IStackSlots__

| Name | Type | Description |
|--------|:------:|-------------|
| `root` | `IHTMLDivSlot` | Defines root slot of the component |
| `inner` | `IHTMLDivSlot` | Defines a slot that is placed inside the root slot in order to achieve wrapping. Only used when the wrap property is set to true. |

__IStackProps__

| Name      | Type     | Default Value | Description |
|-----------|:---------:|:--------------:|-------------|
| `as` | `string` \| `React.ReactType<IStackProps>` |  | Defines how to render the `Stack` |
| `horizontal` | `boolean` | `false` | Defines whether to render `Stack` children horizontally |
| `reversed` | `boolean` | `false` | Defines whether to render `Stack` children in the opposite direction (bottom-to-top if it's a vertical `Stack` and right-to-left if it's a horizontal `Stack`) |
| `horizontalAlign` | `'start'` \| `'flex-start'` \| `'end'` \| `'flex-end'` \| `'center'` \| `'space-between'` \| `'space-around'` \| `'space-evenly'` \| `'baseline'` \| `'stretch'` |  | Defines how to align `Stack` children horizontally (along the x-axis) |
| `verticalAlign` | `'top'` \| `'bottom'` \| `'start'` \| `'flex-start'` \| `'end'` \| `'flex-end'` \| `'center'` \| `'space-between'` \| `'space-around'` \| `'space-evenly'` \| `'baseline'` \| `'stretch'` |  | Defines how to align `Stack` children vertically (along the y-axis) |
| `verticalFill` | `boolean` |  | Defines whether the `Stack` should take up 100% of the height of its parent |
| `disableShrink` | `boolean` |  | Defines whether `Stack` children should not shrink to fit the available space |
| `grow` | `boolean` \| `number` \| `'inherit'` \| `'initial'` \| `'unset'` |  | Defines how much to grow the `Stack` in proportion to its siblings |
| `gap` | `number` \| `string` |  | Defines the spacing between `Stack` children |
| `maxWidth` | `number` \| `string` |  | Defines the maximum width that the `Stack` can take |
| `maxHeight` | `number` \| `string` |  | Defines the maximum height that the `Stack` can take |
| `padding` | `number` \| `string` |  | Defines the inner padding of the `Stack` |
| `wrap` | `boolean` | `false` | Defines whether `Stack` children should wrap onto multiple rows or columns when they are about to overflow the size of the `Stack` |

### Component Ownership

- @kkjeer who initially created the component
- @khmakoto who is promoting the component from `@uifabric/experiments` to `office-ui-fabric-react`

### Deadlines

We are aiming for the component to be promoted from `@uifabric/experiments` to `office-ui-fabric-react` in January 2019.

## Steps

- [x] Themable (Component.base/Component.styles)
- [x] Unit tests
- [x] Visual tests
- [x] Documentation and examples
- [x] Keyboard Accessible
- [x] High Constrast Support
- [x] RTL Support
- [x] Design Review
- [x] API review
- [ ] Public Preview
- [ ] Ready for Publishing

### PR DESCRIPTION
**Pull request by khmakoto**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #5419 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

This PR has the purpose of promoting the **Stack** component from the `experiments` package to the `office-ui-fabric-react` package.

As part of this process, it was decided that we should change the approach we were using in the `experiments` package of having two separate components, **HorizontalStack** and **VerticalStack**, and instead have a single component **Stack** with a boolean property `horizontal` that determines if the stack is horizontal or vertical, with a vertical stack being the default.

This change required modifying existing examples, tests and usage of **Stack** inside other components throughout the `experiments` package. The **HorizontalStack** and **VerticalStack** pages in the `experiments` website have also been removed and a **Stack** page has been added to the `office-ui-fabric-react` website as part of this change.

#### Focus areas to test

All the existing tests (snapshots, vr-tests, etc) for **HorizontalStack** and **VerticalStack** have been modified to use the singular **Stack** component and the new component has been tested against all of them to ensure conformance with how the previous components worked.

#### TO-DOs

- [ ] Add the `wrap` functionality to vertical Stacks.
  - [ ] Add examples in documentation page.
  - [ ] Add snapshot tests.
  - [ ] Add vr-tests.
- [ ] Finish documentation on Stack component.
  - [ ] Add overview.
  - [ ] Add Do's.
  - [ ] Add Dont's.

### CODE REVIEW

File: packages/experiments/src/components/Button/Button.view.tsx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import * as React from 'react';
 import { IButtonComponent, IButtonViewProps } from './Button.types';
 import { Text } from '../../Text';
-import { HorizontalStack } from '../../Stack';
+import { HorizontalStack } from 'office-ui-fabric-react/lib/Stack';
```
Comment (JasonGore) [inline packages/experiments/src/components/Button/Button.view.tsx]: This should just be absorbed into line 5 (don't need to import from lib anymore)

File: packages/experiments/src/components/Button/examples/Button.Example.tsx [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import * as React from 'react';
 import { Button } from '../index';
-import { HorizontalStack, Text, VerticalStack } from '@uifabric/experiments';
+import { HorizontalStack, VerticalStack } from 'office-ui-fabric-react/lib/Stack';
```
Comment (JasonGore) [inline packages/experiments/src/components/Button/examples/Button.Example.tsx]: Also can be moved into line 5. Basically consolidate stack imports into any existing import from OUFR lines.

**Comment by KevinTCoughlin** [comment]
> 
> 
> There are quite a few more `Text` -> `span` conversions than I expected. Since changing those will also affect all of the new snapshots, I'd like to see if Natalie can just piggyback Text promotion onto this branch. That'll also help make sure we don't miss converted any spans back to Text.

FWIW I think that's reasonable 😃

**Comment by natalieethell** [comment]
Just to update the thread here - we spoke offline and may hold off on adding Text to this PR. We may want to work out some things with `ITypography` before promoting Text to OUFR. I'll report back here when I have some more updates.

**Comment by khmakoto** [comment]
@micahgodbolt I mentioned this in the email thread, but just to have it here, I updated the code so that the left/right/center notation is now start/end/center to adhere to LTR.

**Comment by micahgodbolt** [comment]
We also have an issue for stack that it would be great to update/link to https://github.com/OfficeDev/office-ui-fabric-react/issues/5419

**Comment by micahgodbolt** [comment]
@khmakoto i've updated the original issue with or current 'new component' template. Can you move your comments, todos, code examples etc into that issue. 

Another thought is that we should focus on getting Stack correct in Experiments before moving it over to OUFR. i.e. smaller, more focused PRs to prep this control. Then one big PR to simply move it.

**Comment by dzearing** [comment]
Closing, please resolve and reopen when ready.

**Pull request closed without merging**